R/imputation.R

```diff
@@ -272,53 +303,80 @@ imputeMethods <- function()
 ##' @param k `numeric(1)` providing the imputation value used for the
 ##'     first and last samples if they contain an `NA`. The default is
 ##'     to use the smallest value in the data.
-impute_neighbour_average <- function(x, k = min(x, na.rm = TRUE)) {
+impute_neighbour_average <- function(x, k = min(x, na.rm = TRUE), margin = 1L) {
     message("Assuming values are ordered.")
-    .Call("C_impNeighbourAvg", x, k)
+    margin <- .checkMargin(margin)
+    if (margin == 2L) x <- t(x)
+    res <- .Call("C_impNeighbourAvg", x, k)
+    if (margin == 2L) return(t(res))
+    else return(res)
 }
 
 ##' @export
 ##' @rdname imputation
-impute_knn <- function(x, ...) {
+impute_knn <- function(x, margin = 1L, ...) {
     requireNamespace("impute")
+    margin <- .checkMargin(margin)
+    if (margin == 2L)
+        x <- t(x)
     imp_res <- impute::impute.knn(x, ...)
     if (!is.null(imp_res$rng.state)) {
         assign(".Random.seed", imp_res$rng.state, envir = .GlobalEnv)
     } else {
         rm(".Random.seed", envir = .GlobalEnv)
     }
-    imp_res$data
+    if (margin == 2L)
+        return(t(imp_res$data))
+    else imp_res$data
 }
 
 ##' @export
 ##' @rdname imputation
-impute_mle <- function(x, ...) {
+impute_mle <- function(x, margin = 2L, ...) {
     requireNamespace("norm")
+    margin <- .checkMargin(margin)
+    if (margin == 2L)
+        x <- t(x)
     s <- norm::prelim.norm(x)  ## preliminary manipulations
     th <- norm::em.norm(s, ...) ## find the MLE
     seed <- sample(.Machine$integer.max, 1)
     norm::rngseed(seed) ## set random number generator seed
-    norm::imp.norm(s, th, x)  ## impute missing data under the MLE
+    res <- norm::imp.norm(s, th, x)  ## impute missing data under the MLE
+    if (margin == 2L)
+        res <- t(res)
+    res
 }
 
 ##' @export
 ##' @rdname imputation
-impute_bpca <- function(x, ...) {
+impute_bpca <- function(x, margin = 1L, ...) {
     requireNamespace("pcaMethods")
+    margin <- .checkMargin(margin)
+    if (margin == 2L)
+        x <- t(x)
     nSamples <- dim(x)[2]
     .resultBPCA <- pcaMethods::pca(x,
                                    method = "bpca",
                                    nPcs = (nSamples - 1),
                                    verbose = FALSE,
                                    ...)
-    pcaMethods::completeObs(.resultBPCA)
+    res <- pcaMethods::completeObs(.resultBPCA)
+    if (margin == 2L)
+        res <- t(res)
+    res
 }
 
 ##' @export
 ##' @rdname imputation
-impute_RF <- function(x, ...) {
+impute_RF <- function(x, margin = 2L, ...) {
     requireNamespace("missForest")
-    t(missForest::missForest(t(x), ...)$ximp)
+    margin <- .checkMargin(margin)
+    if (margin == 2L)
+        x <- t(x)
+    res <- missForest::missForest(x, ...)$ximp
```

Comment:
I just checked the docs of `missForest`. It assumes:

>The columns correspond to the variables and the rows to the observations.

It then performs random forest in features space (columns for `missForest`, rows for `MsCoreUtils`). So, you should transpose when `margin == 1L`. Not sure whether it makes sense to impute by sample :man_shrugging:

Reply:
Fixed, thank you for checking.